src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs

```diff
@@ -231,36 +241,54 @@ private void BuildAppConfiguration()
             _hostBuilderContext.Configuration = _appConfiguration;
         }
 
-        private void CreateServiceProvider()
+        internal static IServiceCollection CreateServiceCollection(
+            HostBuilderContext hostBuilderContext,
+            HostingEnvironment hostingEnvironment,
+            PhysicalFileProvider defaultFileProvider,
+            IConfiguration appConfiguration,
+            Func<IServiceProvider> serviceProviderGetter)
         {
             var services = new ServiceCollection();
 #pragma warning disable CS0618 // Type or member is obsolete
-            services.AddSingleton<IHostingEnvironment>(_hostingEnvironment);
+            services.AddSingleton<IHostingEnvironment>(hostingEnvironment);
 #pragma warning restore CS0618 // Type or member is obsolete
-            services.AddSingleton<IHostEnvironment>(_hostingEnvironment);
-            services.AddSingleton(_hostBuilderContext);
+            services.AddSingleton<IHostEnvironment>(hostingEnvironment);
+            services.AddSingleton(hostBuilderContext);
             // register configuration as factory to make it dispose with the service provider
-            services.AddSingleton(_ => _appConfiguration);
+            services.AddSingleton(_ => appConfiguration);
 #pragma warning disable CS0618 // Type or member is obsolete
-            services.AddSingleton<IApplicationLifetime>(s => (IApplicationLifetime)s.GetService<IHostApplicationLifetime>());
+            services.AddSingleton(s => (IApplicationLifetime)s.GetService<IHostApplicationLifetime>());
 #pragma warning restore CS0618 // Type or member is obsolete
             services.AddSingleton<IHostApplicationLifetime, ApplicationLifetime>();
 
             AddLifetime(services);
 
             services.AddSingleton<IHost>(_ =>
```

Comment:
Isn't the `_` parameter in this delegate the `IServiceProvider`? Could we use that instead of passing in a `serviceProviderGetter` delegate?

Reply:
I tried to do this initially. In the default case, `_` and the `IServiceProvider` returned by `serviceProviderGetter` are the same object. These can be different for third-party containers though. This caused an issue with Autofac which [I debugged and then promptly forgot about](https://github.com/dotnet/runtime/issues/36060#issuecomment-592284496). Fortunately, @galakt added a solid regression test.

Reply:
Maybe it might be a good idea to add this to a comment? Nothing major, just a thought.